core/src/main/java/org/eclipse/hono/util/CredentialsObject.java

```diff
@@ -0,0 +1,81 @@
+/**
+ * Copyright (c) 2017 Bosch Software Innovations GmbH.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    Bosch Software Innovations GmbH - initial creation
+ */
+package org.eclipse.hono.util;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Encapsulates the credentials information for a device that was found by the get operation of the
+ * <a href="https://www.eclipse.org/hono/api/Credentials-API/">Credentials API</a>.
+ * <p>
+ * Is mapped internally from json representation by jackson-databind.
+ */
+public final class CredentialsObject {
+    @JsonProperty(CredentialsConstants.FIELD_DEVICE_ID)
+    private String deviceId;
+    @JsonProperty(CredentialsConstants.FIELD_TYPE)
+    private String type;
+    @JsonProperty(CredentialsConstants.FIELD_AUTH_ID)
+    private String authId;
+    @JsonProperty(CredentialsConstants.FIELD_ENABLED)
+    private Boolean enabled;
+    /*
+     * Since the format of the secrets field is not determined by the Credentials API, they are best represented as
+     * key-value maps with key and value both of type String.
+     * The further processing of secrets is part of the validator for the specific type.
+     */
+    @JsonProperty(CredentialsConstants.FIELD_SECRETS)
+    private List<Map<String,String>> secrets;
```

Comment:
I agree  that the secrets are the most interesting part, and are the part where the most flexibility is needed. Thus we should be prepared to an arbitrarily complex JSON structure for secrets of a specific type (the type being defined for special cases).
To model this complexity as a String value (which would have to be parsed again by invoking a JSON parser inside a validator) is probably not the best possible solution.

Let me point out 2 things here:

**1.) Better proposal**

I made several attempts to improve the mapping, here is what I prefer now after considering your comment:

`private List<Map<String,Object>> secrets;`

lets jackson-databind create arbitrarily deep LinkedHashMap's

E.g. a structure like 

```
"secrets" : [ 
{    
  "hash-function" : "sha-512", 
  "salt" : "aG9ubw==",
  "complex" : { 
    "keyL1_1" : { 
        "keyL2_1":"valL2_1" }, 
    "keyL1_2" : "valL1_2"
  } 
}
]
```

creates an internal representation of

```
secrets = {ArrayList@3763}  size = 1
 0 = {LinkedHashMap@3772}  size = 3
  0 = {LinkedHashMap$Entry@3775} "hash-function" -> "sha-512"
  1 = {LinkedHashMap$Entry@3776} "salt" -> "aG9ubw=="
  2 = {LinkedHashMap$Entry@3777} "complex" -> " size = 2"
   key = "complex"
   value = {LinkedHashMap@3764}  size = 2
    0 = {LinkedHashMap$Entry@3785} "keyL1_1" -> " size = 1"
     key = "keyL1_1"
     value = {LinkedHashMap@3765}  size = 1
      0 = {LinkedHashMap$Entry@3793} "keyL2_1" -> "valL2_1"
       key = "keyL2_1"
       value = "valL2_1"
    1 = {LinkedHashMap$Entry@3786} "keyL1_2" -> "valL1_2"
     key = "keyL1_2"
     value = "valL1_2"
```

IMHO one can program on this structure equivalently well compared to the JSON parser API.

Instead of having something like 
```
String tenantId = body.getString(MessageHelper.APP_PROPERTY_TENANT_ID);
```
(which might throw a ClassCastException) we have something like 
```
String secretDeepDown = secretDeepDownMap.get("keyL2_1");
```
that also might throw a ClassCastException.

But we operate on mapped objects as far as possible and do not need to access the proper JSON keys for the standardized values.

The keys inside the secrets structure are addressed in point 2:

**2.) additional improvements for a later step:**
Although the problem comes from the intended flexibility regarding the structure of secrets, we could do more for the support of the "standardized" secrets and offer fully defined objects for the types we define in Hono (see "Standard Credential Types" at the end of the credentials API).

This would by now be for type "hashed-password" and "preshared-key".

For these secrets, we could automatically generate a fully mapped POJO again. Then we offer as much comfort to users of the credentials API as we can IMHO:

- we support well all secret types we define
- but still let the secrets part of the API fully open 

The price of self defined secret types are that the self written validators have to deal with the Map's, since only the definer of a new secrets type can know the structure. But this cannot be avoided anyway.

Would be prio2 in my eyes, though.

**Summary**:
I can well understand your comment, and there was necessary improvement, but I still think the mapped object strategy is worth it after not mapping the secrets to one string anymore.
I rather propose that in the future we offer mapped objects to clients with JSON payload where ever possible - especially for payloads that are well defined and not so partly open like the credentials API.

WDYT?